hw/xbox/mcpx_apu.c

```diff
@@ -826,12 +849,38 @@ static void se_frame(void *opaque)
                     NV_PAVS_VOICE_PAR_STATE_ACTIVE_VOICE)) {
                 MCPX_DPRINTF("voice %d not active...!\n", d->regs[current]);
                 fe_method(d, SE2FE_IDLE_VOICE, d->regs[current]);
+            } else {
+                process_voice(d, mixbins, d->regs[current]);
             }
             MCPX_DPRINTF("next voice %d\n", d->regs[next]);
             d->regs[current] = d->regs[next];
         }
     }
 
+#if GENERATE_MIXBIN_BEEP
+    /* Inject some audio to the mixbin for debugging.
+     * Signal is 1500 Hz sine wave, phase shifted by mixbin number. */
+    for (mixbin = 0; mixbin < NUM_MIXBINS; mixbin++) {
+        for (sample = 0; sample < NUM_SAMPLES_PER_FRAME; sample++) {
+            /* Avoid multiple of 1.0 / NUM_SAMPLES_PER_FRAME for phase shift,
+             * or waves cancel out */
+            float offset = mixbin / (float)NUM_SAMPLES_PER_FRAME -
+                           sample / (float)(NUM_SAMPLES_PER_FRAME + 1);
+            float wave = sinf(offset * M_PI * 2.0f);
+            mixbins[mixbin][sample] += wave * 0x3FFFFF;
+        }
+    }
+#endif
+
+    /* Write VP results to the GP DSP MIXBUF */
+    for (mixbin = 0; mixbin < NUM_MIXBINS; mixbin++) {
```

Comment:
It would be nice if we could move the 4k of mix bin data off the stack and write directly into DSP memory here.

Reply:
See original PR description and issue #158.

We also do not only have to do 1024 writes.. it would be more: Each voice (up to 256) will access these mixbins (both read and write; up to 8 times). This access happens in `process_voice` in the future.

By having this temporary array, we avoid costly DSP writes, and only do it once when the final value is determined.

The unused bins seem to be cleared to zero, so we do at least 1024 writes to the DSP memory anyway.
The update is *probably* also atomic at frame start, so we need a temporary buffer.

So if any of these *strong* assumptions (based on memory sniffing on hardware), holds true, then we can't avoid this access; especially if we want multi-threading in the future.